File: benchmark_client/scripts/python/Random Walk.py
import json
import time

import numpy as np

N_WALKERS = 1_000_000
N_STEPS = 1000
N_BINS_2D = 100
# Batched accumulation avoids holding (walkers × steps) matrices for all walkers at once.
_BATCH = 125_000


def run_1d(rng: np.random.Generator) -> tuple[np.ndarray, np.ndarray]:
    sum_pos = np.zeros(N_STEPS, dtype=np.float64)
    sum_pos_sq = np.zeros(N_STEPS, dtype=np.float64)
    hist_len = 2 * N_STEPS + 1
    hist = np.zeros(hist_len, dtype=np.int64)
    start = 0
    while start < N_WALKERS:
        bn = min(_BATCH, N_WALKERS - start)
        rnd = rng.integers(0, 2, size=(bn, N_STEPS), dtype=np.int8)
        moves = np.where(rnd == 0, -1, 1).astype(np.int16, copy=False)
        pos = np.cumsum(moves, axis=1, dtype=np.int32)
        sum_pos += pos.sum(axis=0, dtype=np.float64)
        sum_pos_sq += (pos.astype(np.float64, copy=False) ** 2).sum(axis=0)
        idx = pos[:, -1].astype(np.int64, copy=False) + N_STEPS
        np.add.at(hist, idx, 1)
        start += bn
    return np.stack([sum_pos, sum_pos_sq], axis=0), hist


def run_2d(rng: np.random.Generator) -> tuple[np.ndarray, np.ndarray]:
    sum_r2 = np.zeros(N_STEPS, dtype=np.float64)
    hist = np.zeros(N_BINS_2D, dtype=np.int64)
    max_r = np.sqrt(2.0 * N_STEPS)
    start = 0
    while start < N_WALKERS:
        bn = min(_BATCH, N_WALKERS - start)
        rnd = rng.integers(0, 2, size=(bn, N_STEPS, 2), dtype=np.int8)
        moves_x = np.where(rnd[..., 0] == 0, -1, 1).astype(np.int16, copy=False)
        moves_y = np.where(rnd[..., 1] == 0, -1, 1).astype(np.int16, copy=False)
        x = np.cumsum(moves_x, axis=1, dtype=np.int32)
        y = np.cumsum(moves_y, axis=1, dtype=np.int32)
        x64 = x.astype(np.float64, copy=False)
        y64 = y.astype(np.float64, copy=False)
        sum_r2 += (x64 * x64 + y64 * y64).sum(axis=0)
        xf = x[:, -1].astype(np.float64)
        yf = y[:, -1].astype(np.float64)
        r_final = np.sqrt(xf * xf + yf * yf)
        b = np.minimum((r_final / max_r * N_BINS_2D).astype(np.int64), N_BINS_2D - 1)
        b = np.maximum(b, 0)
        np.add.at(hist, b, 1)
        start += bn
    return sum_r2, hist


def write_sigma_1d(sum_pos: np.ndarray, sum_pos_sq: np.ndarray) -> None:
    mean = sum_pos / N_WALKERS
    mean_sq = sum_pos_sq / N_WALKERS
    var = mean_sq - mean * mean
    var = np.where((var < 0) & (var > -1e-12), 0.0, var)
    sigma = np.sqrt(var)
    steps = np.arange(1, N_STEPS + 1, dtype=np.float64)
    theory = np.sqrt(steps)
    data = np.column_stack([steps, sigma, theory])
    header = "step,sigma_empirical,sigma_theory"
    np.savetxt(
        "sigma_1d.csv",
        data,
        delimiter=",",
        header=header,
        comments="",
        fmt=["%d", "%.10g", "%.10g"],
    )


def write_hist_1d(hist: np.ndarray) -> None:
    positions = np.arange(-N_STEPS, N_STEPS + 1, dtype=np.int64)
    data = np.column_stack([positions, hist])
    np.savetxt(
        "histogram_1d.csv",
        data,
        delimiter=",",
        header="position,count",
        comments="",
        fmt=["%d", "%d"],
    )


def write_sigma_2d(sum_r2: np.ndarray) -> None:
    steps = np.arange(1, N_STEPS + 1, dtype=np.float64)
    emp = np.sqrt(sum_r2 / N_WALKERS)
    theory = np.sqrt(2.0 * steps)
    data = np.column_stack([steps, emp, theory])
    np.savetxt(
        "sigma_2d.csv",
        data,
        delimiter=",",
        header="step,sigma_empirical,sigma_theory",
        comments="",
        fmt=["%d", "%.10g", "%.10g"],
    )


def write_hist_2d(hist: np.ndarray) -> None:
    max_r = np.sqrt(2.0 * N_STEPS)
    b = np.arange(N_BINS_2D)
    rmin = b.astype(np.float64) * max_r / N_BINS_2D
    rmax = (b.astype(np.float64) + 1.0) * max_r / N_BINS_2D
    data = np.column_stack([b, rmin, rmax, hist.astype(np.uint64)])
    np.savetxt(
        "histogram_2d.csv",
        data,
        delimiter=",",
        header="bin,r_min,r_max,count",
        comments="",
        fmt=["%d", "%.5f", "%.5f", "%d"],
    )


def main() -> int:
    rng = np.random.default_rng()

    t0 = time.perf_counter()
    agg_1d, hist_1d = run_1d(rng)
    sum_pos = agg_1d[0]
    sum_pos_sq = agg_1d[1]
    sum_r2, hist_2d = run_2d(rng)
    elapsed_ms = (time.perf_counter() - t0) * 1000.0

    print(json.dumps({"execution_time_ms": elapsed_ms}, separators=(",", ":")))

    write_sigma_1d(sum_pos, sum_pos_sq)
    write_hist_1d(hist_1d)
    write_sigma_2d(sum_r2)
    write_hist_2d(hist_2d)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmark_client/scripts/cpp/Random Walk.cpp
#include <chrono>
#include <cmath>
#include <cstdio>
#include <random>
#include <vector>

namespace {

constexpr std::size_t kWalkers = 1'000'000;
constexpr int kSteps = 1000;
constexpr int kBins2d = 100;

}  // namespace

int main() {
    std::mt19937 rng(std::random_device{}());
    std::uniform_int_distribution<int> step_dist(0, 1);

    std::vector<double> sum_pos(kSteps, 0.0);
    std::vector<double> sum_pos_sq(kSteps, 0.0);
    std::vector<long long> hist_1d(2 * kSteps + 1, 0);

    std::vector<double> sum_r2(kSteps, 0.0);
    std::vector<unsigned long long> hist_2d(kBins2d, 0);

    const auto t0 = std::chrono::high_resolution_clock::now();

    for (std::size_t w = 0; w < kWalkers; ++w) {
        int pos = 0;
        for (int s = 0; s < kSteps; ++s) {
            pos += (step_dist(rng) == 0) ? -1 : 1;
            sum_pos[s] += static_cast<double>(pos);
            sum_pos_sq[s] += static_cast<double>(pos) * static_cast<double>(pos);
        }
        hist_1d[static_cast<std::size_t>(pos + kSteps)] += 1;
    }

    const double max_r = std::sqrt(2.0 * kSteps);

    for (std::size_t w = 0; w < kWalkers; ++w) {
        int x = 0;
        int y = 0;
        for (int s = 0; s < kSteps; ++s) {
            x += (step_dist(rng) == 0) ? -1 : 1;
            y += (step_dist(rng) == 0) ? -1 : 1;
            const double r2 = static_cast<double>(x) * static_cast<double>(x) +
                              static_cast<double>(y) * static_cast<double>(y);
            sum_r2[s] += r2;
        }
        const double r = std::sqrt(static_cast<double>(x) * static_cast<double>(x) +
                                   static_cast<double>(y) * static_cast<double>(y));
        int bin = static_cast<int>(r / max_r * kBins2d);
        if (bin < 0) {
            bin = 0;
        }
        if (bin >= kBins2d) {
            bin = kBins2d - 1;
        }
        hist_2d[static_cast<std::size_t>(bin)] += 1ULL;
    }

    const auto t1 = std::chrono::high_resolution_clock::now();
    const double elapsed_ms =
        std::chrono::duration<double, std::milli>(t1 - t0).count();

    std::printf("{\"execution_time_ms\": %.6f}\n", elapsed_ms);

    FILE *f_sigma = std::fopen("sigma_1d.csv", "w");
    if (f_sigma) {
        std::fprintf(f_sigma, "step,sigma_empirical,sigma_theory\n");
        for (int s = 0; s < kSteps; ++s) {
            const double mean = sum_pos[s] / static_cast<double>(kWalkers);
            const double mean_sq = sum_pos_sq[s] / static_cast<double>(kWalkers);
            double var = mean_sq - mean * mean;
            if (var < 0 && var > -1e-12) {
                var = 0;
            }
            const double sigma = std::sqrt(var);
            std::fprintf(f_sigma, "%d,%.10g,%.10g\n", s + 1, sigma,
                         std::sqrt(static_cast<double>(s + 1)));
        }
        std::fclose(f_sigma);
    }

    FILE *f_hist = std::fopen("histogram_1d.csv", "w");
    if (f_hist) {
        std::fprintf(f_hist, "position,count\n");
        for (int i = 0; i < 2 * kSteps + 1; ++i) {
            std::fprintf(f_hist, "%d,%lld\n", i - kSteps,
                         static_cast<long long>(hist_1d[static_cast<std::size_t>(i)]));
        }
        std::fclose(f_hist);
    }

    FILE *f_sigma2 = std::fopen("sigma_2d.csv", "w");
    if (f_sigma2) {
        std::fprintf(f_sigma2, "step,sigma_empirical,sigma_theory\n");
        for (int s = 0; s < kSteps; ++s) {
            const double mean_r2 = sum_r2[s] / static_cast<double>(kWalkers);
            std::fprintf(f_sigma2, "%d,%.10g,%.10g\n", s + 1, std::sqrt(mean_r2),
                         std::sqrt(2.0 * static_cast<double>(s + 1)));
        }
        std::fclose(f_sigma2);
    }

    FILE *f_hist2 = std::fopen("histogram_2d.csv", "w");
    if (f_hist2) {
        std::fprintf(f_hist2, "bin,r_min,r_max,count\n");
        for (int b = 0; b < kBins2d; ++b) {
            const double rmin = static_cast<double>(b) * max_r / static_cast<double>(kBins2d);
            const double rmax =
                static_cast<double>(b + 1) * max_r / static_cast<double>(kBins2d);
            std::fprintf(f_hist2, "%d,%.5f,%.5f,%llu\n", b, rmin, rmax,
                         static_cast<unsigned long long>(hist_2d[static_cast<std::size_t>(b)]));
        }
        std::fclose(f_hist2);
    }

    return 0;
}
